proposals/memory64/Overview.md

```diff
@@ -2,12 +2,188 @@
 
 ## Summary
 
-This proposal extends WebAssembly with 64-bit indexes to linear memory.
+This page describes a proposal to support linear memory of sizes larger than
+2<sup>32</sup> bits. It provides no new instructions, but instead extends the
+currently existing instructions to allow 64-bit indexes.
 
 ## Motivation
 
-TODO
+[WebAssembly linear memory objects][memory object] have sizes measured in
+[pages][memory page]. Each page is 65536 (2<sup>16</sup>) bytes. In WebAssembly
+version 1, a linear memory can have at most 65536 pages, for a total of
+2<sup>32</sup> bytes (4 [gibibytes][gibibyte]).
+
+In addition to this page limit, all [memory instructions][] currently use the
+[`i32` type][i32] as a memory index. This means they can address at most
+2<sup>32</sup> bytes as well.
+
+For many applications, 4 gibibytes of memory is enough. Using 32-bit memory
+indexes is sufficient in this case, and has the additional benefit that
+pointers in the producer language are smaller, which can yield memory savings.
+However, for applications that need more memory than this, there are no easy
+workarounds given the current WebAssembly feature set. Allowing the WebAssembly
+module to choose between 32-bit and 64-bit memory indexes addresses both
+concerns.
+
+Similarly, since WebAssembly is a Virtual [Instruction Set Architecture][ISA]
+(ISA), some hosts may want to use the WebAssembly binary format as a portable
+executable format, in addition to supporting other non-virtual ISAs. Nearly all
+ISAs have support for 64-bit memory addresses now, and a host may not want to
+have to support 32-bit memory addresses in their ABI.
 
 ## Overview
 
-TODO
+### Structure
+
+* The [limits][syntax limits] structure is changed to use `u64`
+  - `limits ::= {min u64, max u64?}`
+
+* The [memory type][syntax memtype] structure is extended to have an index type
+  - `memtype ::= limits valtype`
```

Comment:
Do we really want to use valtype here? Maybe it's better to define a proper sub category `idxtype ::= i32 | i64`. It seems weird to syntactically allow floats or even references.

Reply:
Works for me, I'll update.